### PR TITLE
ECMA 6 Functions

### PR DESCRIPTION
**Pull request by BigFav**
This adds native calls to some newly created functions (hyperbolic trig functions and cbrt), and full implementations of other ECMA 6 functions (`log2`, `log1p`, and `expm1`).
#235

### CODE REVIEW

**Comment by josdejong** [comment]
Great! I will review your code tomorrow. 

Btw I see the Travis integration tests failed:

```
 ReferenceError: re1p is not defined
      at _log1pComplex (lib/function/arithmetic/log1p.js:91:37)
```

**Comment by BigFav** [comment]
I investigated the errors:

1. `log2` should return the log base 2 of positive numbers: This is caused by the nodejs native function. `Math.log2(8)` simply returns a suboptimal value.
2. `sinh` should return the `sinh` of very small numbers (avoid returning zero): This is also caused by the nodejs native function. `Math.sinh` simply returns suboptimal answers for low numbers; `Math.sinh(1e-50)` returns `0`.
3. The other two errors are a product of the above error, as calculating sine or cosine for a complex number involves `sinh`.

They seemed to fix these implementations in the future, and these problems are not in the most current version.

File: lib/function/arithmetic/log10.js [hunk 1]
```diff
@@ -31,7 +31,9 @@ function factory (type, config, load, typed) {
   var log10 = typed('log10', {
     'number': function (x) {
       if (x >= 0 || config.predictable) {
-        return Math.log(x) / Math.LN10;
+        return (Math.log10)
```
Comment (josdejong) [inline lib/function/arithmetic/log10.js]: This check needs to be done only once on load. For optimization we could move this into a separate function, like:

```js
var _log10number = Math.log10
  ? Math.log10 
  : function (x) {
    return Math.log(x) / Math.LN10;
  };
```

Comment (josdejong) [inline lib/function/arithmetic/log10.js]: Thinking about it: the JavaScript engines will probably optimize this themselves. We could do a simple benchmark test to see if it improves. If not, I prefer the way you implemented it right now, it's better readable.

**Comment by josdejong** [comment]
Thanks Favian, your work looks good. 

I'm only in doubt about these issues in node.js v0.12. It's too early to drop support for v0.12, and we cannot assume this will be fixed very soon. Maybe it's best to revert usage of the native functions giving problems in v0.12, and put a comment in the code to fix this in the future?

File: lib/function/arithmetic/log10.js (hunk 1 above)
Comment (BigFav) [inline lib/function/arithmetic/log10.js]: I'll test this out tonight.

Comment (BigFav) [inline lib/function/arithmetic/log10.js]: This way is much faster.

Comment (BigFav) [inline lib/function/arithmetic/log10.js]: It is pretty much the same speed in Node.js, Firefox, and Chrome.

Comment (josdejong) [inline lib/function/arithmetic/log10.js]: Thanks for testing, so we can keep this as it is, great :)

**Comment by josdejong** [comment]
@BigFav I think it's acceptable to have these round-off errors (knowing that with the newest versions of node.js it's improved). What do you think?

**Comment by josdejong** [comment]
@BigFav I see this PR is still open. If I remember well we can merge it and there are no open issues with it, right?

**Comment by josdejong** [comment]
(except for round-off errors on node.js 0.12, we should adjust the unit tests for that I guess)

**Comment by infusion** [comment]
What about a global file which fixes all these ES6 checks at once and for every occurrence in the code, Math.*** can be used quite naturally.

**Comment by josdejong** [comment]
You mean create a polyfill for these functions, which create polyfills of the functions only when missing? Makes sense.

**Comment by infusion** [comment]
Yep, exaclty. This way the polyfills are applied at the init phase (page load, script start) and not for every call.

**Comment by harrysarson** [comment]
Given that mathjs has dropped support for node v0.x can this now be merged?

**Comment by josdejong** [comment]
wow, good point! Yes I guess we can see whether we can merge this PR now. Maybe you can merge it into a separate feature branch on mathjs so you can fix the merge conflicts.

File: lib/function/arithmetic/log10.js (hunk 1 above)
Comment (infusion) [reply lib/function/arithmetic/log10.js]: JS engines do not optimize these functions yet. They are implemented as plain JavaScript, which is the reason for equivalent performance.

**Comment by harrysarson** [comment]
All changes in 746233d are now in mathjs. @BigFav if you could revert 746233d and then I think that adb61b53511 (new es6 methods) can be merged without conflicts.

**Comment by josdejong** [comment]
@harrysarson I don't think Favian is still active on the project (at least not this period), so I think we should merge his work into a new feature branch in mathjs and solve the merge conflicts ourselves.

**Comment by harrysarson** [comment]
See #1046.

**Pull request closed without merging**